File: 240_search_2d_matrix_ii/m+n.py
class Solution:
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        """
        Complexity: O(m+n)
        Space: O(1)
        Started from the top-right, if the target is smaller, move left (one col)
        else move down (one row)
        """
        if matrix == [] or matrix[0] == []:
            return False
        m = 0
        n = len(matrix[0]) - 1
        while(m < len(matrix) and n >= 0):
            if matrix[m][n] == target:
                return True
            if matrix[m][n] > target:
                n -= 1
            else:
                m += 1
                
        return False

File: 074_search_a_2d_matrix/m+n.cpp
class Solution {
  /*
   * Complexity: O(m+n)
   * Space: O(1)
   */
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        if(!m) return false;
        int n = matrix[0].size();
        if(!n) return false;
        for(int row = 0; row < m; ++row){
            if(matrix[row][n-1] > target){
                for(int col = 0; col < n; ++col){
                    if(matrix[row][col] == target) return true;
                }
                return false;
            }
            else if(matrix[row][n-1] == target) return true;
        }
        return false;
    }
};
